core/templates/services/messenger.service.ts

```diff
@@ -19,147 +19,196 @@
+
+  SUPPORTED_HASHDICT_VERSIONS: Set<string> = (
+    new Set(['0.0.0', '0.0.1', '0.0.2', '0.0.3']));
+
+  MESSAGE_VALIDATORS: IMessageValidatorsType = {
+    heightChange(payload: IHeightChangeData): boolean {
+      const {height, scroll} = payload;
+      return (
+        Number.isInteger(height) && height > 0 && typeof scroll === 'boolean');
+    },
+    explorationLoaded(): boolean {
+      return true;
+    },
+    stateTransition(payload: IStateTransitionData): boolean {
+      return Boolean(payload.oldStateName) || Boolean(payload.newStateName);
+    },
+    explorationReset(payload: {stateName: string}): boolean {
+      return Boolean(payload.stateName);
+    },
+    explorationCompleted(): boolean {
+      return true;
+    }
+  };
+
+  getPayload: GetPayloadType = {
+    heightChange(data: IHeightChangeData): IHeightChangeData {
+      return {
+        height: data.height,
+        scroll: data.scroll
+      };
+    },
+    explorationLoaded(data: IExplorationLoadedData): IExplorationLoadedData {
+      return {
+        explorationVersion: data.explorationVersion,
+        explorationTitle: data.explorationTitle
+      };
+    },
+    stateTransition(data: IStateTransitionData): IStateTransitionData {
+      return {
+        explorationVersion: data.explorationVersion,
+        oldStateName: data.oldStateName,
+        jsonAnswer: data.jsonAnswer,
+        newStateName: data.newStateName
+      };
+    },
+    explorationCompleted(
+        data: IExplorationCompletedData): IExplorationCompletedData {
+      return {
+        explorationVersion: data.explorationVersion
+      };
+    },
+    // DEPRECATED
+    explorationReset(data: string): {stateName: string} {
+      return {
+        stateName: data
+      };
+    }
+  };
+
+  // The argument messageData is of type Object since it can have
+  // different properties based on the messageTitle.
+  sendMessage(messageTitle: string, messageData: Object): void {
+    // TODO(sll): For the stateTransition and explorationCompleted events,
+    // we now send paramValues in the messageData. We should broadcast these
+    // to the parent page as well.
+    // TODO(sll): Delete/deprecate 'reset exploration' from the list of
+    // events sent to a container page.
+
+    // Only send a message to the parent if the oppia window is iframed and
+    // a hash is passed in.
+    var window = this.windowRef.nativeWindow;
+    var rawHash = window.location.hash.substring(1);
+    if (window.parent !== window && rawHash &&
+        this.MESSAGE_VALIDATORS.hasOwnProperty(messageTitle)) {
+      // Protractor tests may prepend a / to this hash, which we remove:
+      var hash =
+        (rawHash.charAt(0) === '/') ? rawHash.substring(1) : rawHash;
+      var hashParts = hash.split('&');
+      var hashDict = {
+        version: null,
+        secret: null,
+        tagid: null
+      };
+      for (var i = 0; i < hashParts.length; i++) {
```

Comment:
Can this `var` be changed to `const` or `let`?

Reply:
Done.